api/src/Serializer/Normalizer/UserNormalizer.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace App\Serializer\Normalizer;
+
+use App\Entity\User;
+use App\Metadata\Resource\Factory\UriTemplateFactory;
+use Rize\UriTemplate;
+use Symfony\Component\Serializer\Exception\ExceptionInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+use Symfony\Component\Serializer\SerializerAwareInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+
+class UserNormalizer implements NormalizerInterface, SerializerAwareInterface {
+    public function __construct(
+        private NormalizerInterface $decorated,
+        private UriTemplate $uriTemplate,
+        private UriTemplateFactory $uriTemplateFactory,
+    ) {
+    }
+
+    public function supportsNormalization($data, string $format = null) {
+        return $this->decorated->supportsNormalization($data, $format);
+    }
+
+    /**
+     * @throws ExceptionInterface
+     * @throws \ApiPlatform\Core\Exception\ResourceClassNotFoundException
+     */
+    public function normalize($object, string $format = null, array $context = []) {
+        $data = $this->decorated->normalize($object, $format, $context);
+        if ($object instanceof User) {
+            // get uri for the respective ContentNode entity and add ContentType as query parameter
+            [$uriTemplate] = $this->uriTemplateFactory->createFromResourceClass(User::class);
+            $profileTemplate = str_replace('/users', '/profiles', $uriTemplate);
+            $profileUri = $this->uriTemplate->expand($profileTemplate, ['id' => $object->getId()]);
+            $userUri = $this->uriTemplate->expand($uriTemplate, ['id' => $object->getId()]);
+
+            $data['_links']['self']['href'] = $userUri;
```

Comment:
Schön wäre, wenn wir zu Profile nicht nur den Link haben, sondern die Daten auch embedden können. Wenn wir eine CampCollaboration Liste von 20 Leitern haben setzten wir ansonsten 20 Einzelrequests ab, um die Details zu laden.

Wollen wir User und Profile nicht einfach als separate Entities modellieren? Dann kommt das halt alles gratis mit.

Reply:
Brauchen wir momentan (halt nur momentan) nicht, da wir auf Collaborator.vue nur den DisplayName vom User anzeigen. (Aber die Migration in der Produktion wäre möglich, wahrscheinlich schon nicht so toll).

Da ist dann wieder die Frage:

1. Seperater Endpoint für /profile, oder "subresource" (nur embedded im User falls berechtigung da)
2. Link/Embedding nur Einfügen, wenn die Berechtigung da ist?
2. Als abstrakte Klasse oder als 1:1 Relation?
3. Wenn 1:1 Relation, sowohl bei User als auch bei Profile den Link einbauen?
4. Für jede Berechtigungstufe eine eigene Klasse machen? Welche Klassen verlinken/embedden zu welchen?

Reply:
Ich sehe es glaube ich ähnlich wie usu, habe aber noch nicht alles 100% durchgedacht. Ich versuche mal, deine Fragen aus meiner Sicht zu beantworten.
1. Separate Entities (oder separate ApiResources mit gemeinsamer Entity), bedeuten für mich automatisch separate Endpoints
2. Links würde ich immer einfügen, da es im Frontend nicht trivial ist, herauszufinden ob ein Link wirklich nicht existiert (während loading oder reloading kann man das nicht wissen, und ein reload kann auch während einem delete einer anderen Entity getriggert werden, etc...). Embedding natürlich nur wenn die Berechtigung da ist.
3. Mein Vorschlag wäre eine gemeinsame Tabelle, gemeinsame Doctrine Entity, aber mehrere ApiResources die dieselbe Entity verwenden. Das ist möglich via Custom [DataProvider](https://api-platform.com/docs/core/data-providers/) (lesend) und [DataPersister](https://api-platform.com/docs/core/data-persisters/) (schreibend), aber hier bin ich eben noch nicht sicher ob das die einfachste Lösung ist (funktionieren dann z.B. ApiFilter out of the box?). Wenns mit Vererbung einfacher geht, why not. Eine Relation mit komplett separaten Tabellen wäre eher ein last resort für mich, wegen der zusätzlichen Komplexität im Datenmodell.
4. Am liebsten würde ich der Bequemlichkeit halber von jeder Variante auf jede Variante verlinken. Es könnte sein dass das dann ab 3 Varianten zu unendlich vielen JOINs führt in der EagerLoadingExtension. Falls das der Fall ist würde ich die "öffentliche" Variante als Einstiegspunkt nehmen und von dort auf alle anderen Varianten verlinken.
5. Dito 4.

Reply:
## Core Meeting

Wir gehen mit der Variante: "Separierte Entitäten mit 1:1 Relation"
Es gibt immer beide.

Reply:
done